src/libraries/Microsoft.Extensions.Options/gen/Parser.cs

```diff
@@ -667,12 +667,20 @@ private static string EscapeString(string s)
 
         private void Diag(DiagnosticDescriptor desc, Location? location)
         {
-            _reportDiagnostic(Diagnostic.Create(desc, location, Array.Empty<object?>()));
+            // ignore the diagnostics outside the current compilation
+            if (location?.Kind == LocationKind.SourceFile)
+            {
+                _reportDiagnostic(Diagnostic.Create(desc, location, Array.Empty<object?>()));
+            }
         }
 
         private void Diag(DiagnosticDescriptor desc, Location? location, params object?[]? messageArgs)
         {
-            _reportDiagnostic(Diagnostic.Create(desc, location, messageArgs));
+            // ignore the diagnostics outside the current compilation
+            if (location?.Kind == LocationKind.SourceFile)
```

Comment:
That covers only half of cases - when an assembly is a 3rd party one. We can still consider producing diagnostics in that case, but the location needs to be some fallback one.

When the assembly is yours (i.e. you have its source code), that condition will be `true`, but VS plugins will crash.

Reply:
I believe the System.Text.Json source generator may be impacted by the same issue (although we haven't received any customer reports about this). What is your recommendation when it comes to picking a fall back location? Just setting to `null`?

cc @layomia who might be interested in this.

Reply:
We can consider making an `OptionsValidator` to be a fallback in that case (worst case scenario is when a validator is applied on a type residing in another assembly).
For JSON source-gen, a class annotated with `[JsonSerializable]` can be considered as a fallback location.

Reply:
In both cases types annotated with an attribute that kicks in source-generation should definitely belong to the compilation unit in which a source-gen is executing.

Reply:
Thanks, I'll run an audit of the implementation to see whether fallback is necessary.

Reply:
> When the assembly is yours (i.e. you have its source code), that condition will be true, but VS plugins will crash.

Are we sure of that? the compilation will be referencing your assembly as a meta data at that time. no?

Reply:
I was debugging another generator locally and it was the case for me. You can check your fix by running it on the code I provided in #88152.